File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/interpreter/solution/MethodInvocation.java
// MethodInvocation.java

import java.io.PrintWriter;

public class MethodInvocation extends Nonterminal
{
	public MethodInvocation(String aName)
	{
		setName(aName);
		myOperands = new DpVector<Expression>();
	}

	public int getNumberOfUsesOfVariable(String aName)
	{
		int result = 0;

		DpIterator<Expression> aIterator = getOperands();
		while (!aIterator.isDone())
		{
			Expression aOperand = aIterator.getCurrent();
			if (aOperand == null)
			{
				throw new DpException("int MethodInvocation.getNumberOfUsesOfVariable(String aName) -- aOperand == null.");
			}
			result += aOperand.getNumberOfUsesOfVariable(aName);
			aIterator.advance();
		}

		return result;
	}

	public double getValue(DpMap<String, Double> aMap)
	{
		double result = 0;

		if (getName().equals("sqrt"))
		{
			DpIterator<Expression> aIterator = getOperands();
			Expression aOperand1 = aIterator.getCurrent();
			if (aOperand1 == null)
			{
				throw new DpException("double MethodInvocation.getValue(DpMap<String, Double> aMap) -- aOperand1 == null.");
			}
			double aValue1 = aOperand1.getValue(aMap);

			result = Math.sqrt(aValue1);
		}
		else
		{
			throw new DpException("double MethodInvocation.getValue(DpMap<String, Double> aMap) -- unsupported method name.");
		}

		return result;
	}

	public void output(PrintWriter aPrintWriter)
	{
		aPrintWriter.print(getName());
		aPrintWriter.print("(");
		boolean isFirst = true;
		DpIterator<Expression> aIterator = getOperands();
		while (!aIterator.isDone())
		{
			if (isFirst)
			{
				isFirst = false;
			}
			else
			{
				aPrintWriter.print(",");
			}
			Expression aOperand1 = aIterator.getCurrent();
			if (aOperand1 == null)
			{
				throw new DpException("void MethodInvocation.output(PrintWriter aPrintWriter) -- aOperand1 == null.");
			}
			aOperand1.output(aPrintWriter);
			aIterator.advance();
		}
		aPrintWriter.print(")");
	}

	private void setName(String aName)
	{
		myName = aName;
	}

	public String getName()
	{
		return myName;
	}

	public void addOperand(Expression aOperand)
	{
		myOperands.add(aOperand);
	}

	public DpIterator<Expression> getOperands()
	{
		return myOperands.makeIterator();
	}

	private String myName;
	private DpVector<Expression> myOperands;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/interpreter/solution/MethodInvocation.m
% MethodInvocation.m

classdef MethodInvocation < Nonterminal
	properties (GetAccess = public, SetAccess = private)
		Name % string
		Operands % DpVector of Expression
	end

	methods
		% constructor:
		function theMethodInvocation = MethodInvocation(aName)
			theMethodInvocation.Name = aName;
			theMethodInvocation.Operands = DpVector;
		end

		function result = getNumberOfUsesOfVariable(theMethodInvocation, aName)
			result = 0;
			aIterator = theMethodInvocation.getOperands;
			while ~aIterator.isDone
				aOperand = aIterator.getCurrent;
				if isempty(aOperand)
					error('aOperand is empty.');
				end
				result = result + aOperand.getNumberOfUsesOfVariable(aName);
				aIterator.advance;
			end
		end

		function result = getValue(theMethodInvocation, aMap)
			if strcmp(theMethodInvocation.Name, 'sqrt')
				aIterator = theMethodInvocation.getOperands;
				aOperand1 = aIterator.getCurrent;
				if isempty(aOperand1)
					error('aOperand1 is empty.');
				end
				aValue1 = aOperand1.getValue(aMap);

				result = sqrt(aValue1);
			else
				error('Unsupported method Name.');
			end
		end

		function output(theMethodInvocation)
			fprintf('%s', theMethodInvocation.Name);
			fprintf('(');
			isFirst = true;
			aIterator = theMethodInvocation.getOperands;
			while ~aIterator.isDone
				if isFirst
					isFirst = false;
				else
					fprintf(',');
				end
				aOperand = aIterator.getCurrent;
				if isempty(aOperand)
					error('aOperand is empty.');
				end
				aOperand.output;
				aIterator.advance;
			end
			fprintf(')');
		end

		function addOperand(theMethodInvocation, aOperand)
			theMethodInvocation.Operands.add(aOperand);
		end

		function result = getOperands(theMethodInvocation)
			result = theMethodInvocation.Operands.makeIterator;
		end
	end
end
